### PR TITLE
⚠️ Use k8s v1.27.0 in quickstart docs and CAPD

### PR DESCRIPTION
**Pull request by furkatgofurov7**
Signed-off by: Furkat Gofurov (furkat.gofurov@suse.com)

**What this PR does / why we need it**:
Bumps k8s version to v1.27.0 in Quickstart docs and CAPD. 

**NOTE**: 

- etcd version is bumped to` 3.5.7-0` following the [kubeadm](https://github.com/kubernetes/kubernetes/blob/release-1.27/cmd/kubeadm/app/constants/constants.go#L311-L312)
-  COREDNS_VERSION is bumped to `v1.10.1` following the [kubeadm](https://github.com/kubernetes/kubernetes/blob/c3e3ff989ac9abfb0dc3b958fcb698fcce952f8c/cmd/kubeadm/app/constants/constants.go#L348)
- new field called: `WorkloadKubernetesVersion` was introduced to `ClusterctlUpgradeSpecInput`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Part of: #8459 

cc @kubernetes-sigs/cluster-api-release-team

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
/hold

Manually verifying quickstart is pending

**Comment by furkatgofurov7** [comment]
> /hold
> 
> Manually verifying quickstart is pending

Done and all good.

/hold cancel

File: test/e2e/config/docker.yaml [hunk 1]
```diff
@@ -323,11 +323,11 @@ variables:
...
   # The following Kubernetes versions should be the latest versions with already published kindest/node images.
   # This avoids building node images in the default case which improves the test duration significantly.
-  KUBERNETES_VERSION_MANAGEMENT: "v1.26.0"
-  KUBERNETES_VERSION: "v1.26.0"
-  KUBERNETES_VERSION_UPGRADE_FROM: "v1.25.3"
-  KUBERNETES_VERSION_UPGRADE_TO: "v1.26.0"
-  ETCD_VERSION_UPGRADE_TO: "3.5.6-0"
+  KUBERNETES_VERSION_MANAGEMENT: "v1.27.0"
+  KUBERNETES_VERSION: "v1.27.0"
+  KUBERNETES_VERSION_UPGRADE_FROM: "v1.26.0"
```
Comment (joekr) [inline test/e2e/config/docker.yaml:328]: Should this be the [latest](https://kubernetes.io/releases/) of `1.26.3` like previous?

File: test/e2e/clusterctl_upgrade_test.go [hunk 2]
```diff
@@ -143,7 +143,7 @@ var _ = Describe("When testing clusterctl upgrades (v1.3=>current)", func() {
 			// try to deploy the latest version of our test-extension from docker.yaml.
 			InitWithRuntimeExtensionProviders: []string{},
 			InitWithProvidersContract:         "v1beta1",
-			InitWithKubernetesVersion:         "v1.26.0",
+			InitWithKubernetesVersion:         "v1.27.0",
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:146]:
```suggestion
			InitWithKubernetesVersion:         "v1.26.0",
```

The maximum support version for v1.3 will remain v1.26, so let's keep this on v1.26

File: test/e2e/clusterctl_upgrade_test.go [hunk 3]
```diff
@@ -181,7 +181,7 @@ var _ = Describe("When testing clusterctl upgrades using ClusterClass (v1.3=>cur
...
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:184]:
```suggestion
			InitWithKubernetesVersion:         "v1.26.0",
```
Same

File: test/e2e/clusterctl_upgrade_test.go [hunk 4]
```diff
@@ -209,7 +209,7 @@ var _ = Describe("When testing clusterctl upgrades (v1.4=>current)", func() {
 			SkipCleanup:               skipCleanup,
 			InitWithBinary:            "https://github.com/kubernetes-sigs/cluster-api/releases/download/v1.4.0/clusterctl-{OS}-{ARCH}",
 			InitWithProvidersContract: "v1beta1",
-			InitWithKubernetesVersion: "v1.26.0",
+			InitWithKubernetesVersion: "v1.27.0",
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:212]: Technically v1.4 will only support running on v1.27 after the first v1.4.x release which supports v1.27 but let's just run the full tests on this PR and then bump on here already.

tl;dr let's run e2e-full on this PR and then this change is fine

**Review by sbueringer — COMMENTED** [review]
Just a few nits

**Comment by sbueringer** [comment]
@furkatgofurov7 @killianmuldoon Let's wait with test-infra PR merges until this PR is green

**Comment by killianmuldoon** [comment]
Definitely consistent failure in the 0.3 upgrade test :thinking:. I'll have a look certainly by tomorrow some time (could have been a good candidate for the hacking session :smile:)

File: test/e2e/clusterctl_upgrade_test.go (hunk 3 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade_test.go]: Done

File: test/e2e/clusterctl_upgrade_test.go (hunk 2 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade_test.go]: Oh yeah, thx, great catch, updated now.

**Comment by sbueringer** [comment]
Ups, is this test actually trying to create a 1.27 workload cluster with CAPI v0.3?


If I see this correctly we should introduce a new parameter to the test spec for the test cluster kubernetes version

**Comment by sbueringer** [comment]
Context:
* This test creates a separate mangement cluster with Kubernetes version: `InitWithKubernetesVersion` 
* Then the old providers are deployed on this cluster
* Then we create a test workload cluster via the old providers

Proposal:
* I would suggest to introduce a new field to `ClusterctlUpgradeSpecInput` called: `WorkloadKubernetesVersion`
* This field should be set to the maximumg workload cluster version supported by the old providers

So we end up with the following:

| Test | InitWithKubernetesVersion | WorkloadKubernetesVersion |
|------|---------------------------|---------------------------|
| v0.3 | v1.21.14                  | v1.22.17                  |
| v0.4 | v1.23.17                  | v1.23.17                  |
| v1.0 | v1.23.17                  | v1.23.17                  |
| v1.3 | v1.26.3                   | v1.26.3                   |
| v1.4 | v1.27.0                   | v1.27.0                   |

InitWithKubernetesVersion is the version we deploy the old provider on
WorkloadKubernetesVersion is the version of the workload cluster the old provider creates

Both those version should be the highest the old provider supports

Note: patch version should be the highest we have corresponding kindest/node image for on dockerhub. I looked them up and "bumped" them a bit. We don't have to continuously keep them up-to-date, but maybe let's do it now as we have to introduce WorkloadKubernetesVersion anyway

@furkatgofurov7  @killianmuldoon @fabriziopandini @ykakarap Opinions?

**Comment by furkatgofurov7** [comment]
> Ups, is this test actually trying to create a 1.27 workload cluster with CAPI v0.3?

yeah, but that is the case with 0.4, as well no? And I saw that went fine 🤔

**Comment by sbueringer** [comment]
> > Ups, is this test actually trying to create a 1.27 workload cluster with CAPI v0.3?
> 
> yeah, but that is the case with 0.4, as well no? And I saw that went fine thinking

I could imagine that v0.4 is for some reason compatible with v1.27 (e.g. we use a newer client-go, controller-runtime version).

But it should be relatively easy to see by running the failing test locally and just looking why the Machines are not coming up (I just don't have the time for it right now).

I think we should definitely do it though, just to understand why it is actually failing.

I would do the [suggested change](https://github.com/kubernetes-sigs/cluster-api/pull/8518#issuecomment-1508064859) in any case btw, even if we don't need it for this specific PR. It might just be the solution we need if we find out that v0.3 is simply not compatible with Kubernetes v1.27

**Comment by fabriziopandini** [comment]
+1 to pin the Kubernetes versions we are using when starting the upgrade tests to the latest supported version, this makes sense from a support matrix PoV and it could also improve our test stability

**Comment by furkatgofurov7** [comment]
> I would do the [suggested change](https://github.com/kubernetes-sigs/cluster-api/pull/8518#issuecomment-1508064859) in any case btw, even if we don't need it for this specific PR. It might just be the solution we need if we find out that v0.3 is simply not compatible with Kubernetes v1.27

Sure, I was checking it and noticed all relevant fields starts with `InitWith` prefix, so I went ahead and introduced a new field: `InitWithWorkloadKubernetesVersion` (please check https://github.com/kubernetes-sigs/cluster-api/pull/8518/commits/fb0d73686c3b5d831a41923cca5900e4db4c4d22) to keep them aligned, please let me know if you think otherwise.

**Comment by killianmuldoon** [comment]
Managed to reproduce this. CAPD doesn't come up on the workload cluster with the message:


> your configuration file uses an old API spec: \"kubeadm.k8s.io/v1beta2\". Please use kubeadm v1.22 instead and run 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.\nTo see the stack trace of this error execute with --v=5 or higher\n" "stdout"=

The issue is that the v1beta2 Kubeadm API is no longer supported at all in Kubernetes 1.27+.

Agree completely with @sbueringer that the correct approach is to pin the workload cluster version.

File: test/e2e/clusterctl_upgrade.go [hunk 5]
```diff
@@ -71,6 +72,9 @@ type ClusterctlUpgradeSpecInput struct {
 	// InitWithKubernetesVersion can be used to override the INIT_WITH_KUBERNETES_VERSION e2e config variable with a specific
 	// Kubernetes version to use to create the secondary management cluster, e.g. `v1.25.0`
 	InitWithKubernetesVersion string
+	// InitWithWorkloadKubernetesVersion can be used to override the INIT_WITH_WORKLOAD_KUBERNETES_VERSION e2e config variable with a specific
+	// Kubernetes version to use to create the workload cluster, e.g. `v1.25.0`
+	InitWithWorkloadKubernetesVersion string
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:77]: The idea behind the init suffix is that those are fields related to the clusterctl init call when we install the old provider version (init vs that we upgrade them later).

I think this doesn't work well with the Kubenetes version of the workload cluster that is created with those old provider and later reconciled by the new providers. That's why I suggested WorkloadKubernetesVersion to align to WorkloadFlavor which is what we use to specify the flavor of this same cluster.

Comment (sbueringer) [reply test/e2e/clusterctl_upgrade.go:77]: So for the field I would definitely prefer WorkloadKubernetesVersion. The env variables are a bit of problem as the existing ones are not scoped to this test at all.

I would suggest to expose it via env var as "CLUSTERCTL_UPGRADE_WORKLOAD_KUBERNETES_VERSION". This can be also cherry-picked to the v1.4 branch the same way. I think then it would be good to have a follow-up PR to cleanup the other env vars and also add it to the migration guide (today those env vars are not used by core CAPI, just by other infra providers)

File: test/e2e/clusterctl_upgrade.go [hunk 6]
```diff
@@ -328,12 +340,12 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		// so we are getting a template using the downloaded version of clusterctl, applying it, and wait for machines to be provisioned.
 
 		workLoadClusterName = fmt.Sprintf("%s-%s", specName, util.RandomString(6))
-		kubernetesVersion := input.E2EConfig.GetVariable(KubernetesVersion)
+		workLoadKubernetesVersion := input.E2EConfig.GetVariable(WorkloadKubernetesVersion)
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:343]: This seems to be an additional variable and the one from above is not used?

File: test/framework/clusterctl/client.go [hunk 7]
```diff
@@ -235,17 +235,18 @@ func Delete(_ context.Context, input DeleteInput) {
...
-	Flavor                   string
-	ClusterctlVariables      map[string]string
+	LogFolder                 string
+	ClusterctlConfigPath      string
+	KubeconfigPath            string
+	InfrastructureProvider    string
+	Namespace                 string
+	ClusterName               string
+	KubernetesVersion         string
+	WorkloadKubernetesVersion string
```
Comment (sbueringer) [inline test/framework/clusterctl/client.go:245]:
```suggestion
```
probably not intentional

File: test/framework/clusterctl/clusterctl_helpers.go [hunk 8]
```diff
@@ -273,7 +273,7 @@ func ApplyClusterTemplateAndWait(ctx context.Context, input ApplyClusterTemplate
 	Expect(input.ConfigCluster.WorkerMachineCount).ToNot(BeNil())
 
 	log.Logf("Creating the workload cluster with name %q using the %q template (Kubernetes %s, %d control-plane machines, %d worker machines)",
-		input.ConfigCluster.ClusterName, valueOrDefault(input.ConfigCluster.Flavor), input.ConfigCluster.KubernetesVersion, *input.ConfigCluster.ControlPlaneMachineCount, *input.ConfigCluster.WorkerMachineCount)
+		input.ConfigCluster.ClusterName, valueOrDefault(input.ConfigCluster.Flavor), input.ConfigCluster.WorkloadKubernetesVersion, *input.ConfigCluster.ControlPlaneMachineCount, *input.ConfigCluster.WorkerMachineCount)
```
Comment (sbueringer) [inline test/framework/clusterctl/clusterctl_helpers.go:276]:
```suggestion
		input.ConfigCluster.ClusterName, valueOrDefault(input.ConfigCluster.Flavor), input.ConfigCluster.KubernetesVersion, *input.ConfigCluster.ControlPlaneMachineCount, *input.ConfigCluster.WorkerMachineCount)
```
was that intentional?

File: test/e2e/clusterctl_upgrade.go (hunk 6 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade.go]: Sorry, it was messed up locally and pushed here as well, removed it

File: test/framework/clusterctl/client.go (hunk 7 above)
Comment (furkatgofurov7) [reply test/framework/clusterctl/client.go]: No, removed

File: test/e2e/clusterctl_upgrade.go [hunk 9]
```diff
@@ -71,6 +71,8 @@ type ClusterctlUpgradeSpecInput struct {
 	// InitWithKubernetesVersion can be used to override the INIT_WITH_KUBERNETES_VERSION e2e config variable with a specific
 	// Kubernetes version to use to create the secondary management cluster, e.g. `v1.25.0`
 	InitWithKubernetesVersion string
+	// WorkloadKubernetesVersion is Kubernetes version used to create the workload cluster, e.g. `v1.25.0`
+	WorkloadKubernetesVersion string
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:75]: nit: let's move it down to the WorkloadFlavor

File: test/framework/clusterctl/clusterctl_helpers.go (hunk 8 above)
Comment (furkatgofurov7) [reply test/framework/clusterctl/clusterctl_helpers.go]: No, removed

File: test/e2e/clusterctl_upgrade.go (hunk 9 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade.go:75]: My idea was (missed a detail when writing it up) to use this field if it is set and otherwise fallback to the KUBERNETES_VERSION env var (~ l.333)

Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade.go:75]: @sbueringer ah, I think you mean something like:

```
kubernetesVersion := ""
if input.WorkloadKubernetesVersion != "" {
	kubernetesVersion = input.WorkloadKubernetesVersion
} else {
	kubernetesVersion = input.E2EConfig.GetVariable(KubernetesVersion)
}
```

Comment (sbueringer) [reply test/e2e/clusterctl_upgrade.go:75]: To avoid confusion. In addition to that I meant to move the WorkloadKubenetesVersion down in the struct to the WorkloadFlavor

Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade.go]: Yeap, thanks and done.

**Comment by furkatgofurov7** [comment]
Let's try it once again:


/test pull-cluster-api-e2e-full-main

**Comment by furkatgofurov7** [comment]
> Edit: I will try to repro it on my end

To confirm on this, also seeing the same as below reported by @killianmuldoon:

> > your configuration file uses an old API spec: "kubeadm.k8s.io/v1beta2". Please use kubeadm v1.22 instead and run 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.\nTo see the stack trace of this error execute with --v=5 or higher\n" "stdout"=
> 

> The issue is that the v1beta2 Kubeadm API is no longer supported at all in Kubernetes 1.27+.

and yes, correct, API removal PR is here https://github.com/kubernetes/kubernetes/pull/114540

File: test/e2e/clusterctl_upgrade_test.go [hunk 10]
```diff
@@ -36,7 +36,8 @@ var _ = Describe("When testing clusterctl upgrades (v0.3=>current)", func() {
 			InitWithBinary:            "https://github.com/kubernetes-sigs/cluster-api/releases/download/v0.3.25/clusterctl-{OS}-{ARCH}",
 			InitWithProvidersContract: "v1alpha3",
 			// CAPI v0.3.x does not work on Kubernetes >= v1.22.
-			InitWithKubernetesVersion: "v1.21.12",
+			InitWithKubernetesVersion: "v1.21.14",
+			WorkloadKubernetesVersion: "v1.22.17",
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade_test.go:40]: Very last comment: let's add a note to the v1.4=>v1.5 doc to tell providers that they should do the same
(in fact if they run v0.3 clusterctl upgrade tests they **have** to do the same)

Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade_test.go:40]: yes let's do it, but first i want to let the current test to finish (also, I wonder if there are providers still running 0.3=> current😄)

File: test/e2e/clusterctl_upgrade.go (hunk 5 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade.go]: I will open a separate issue to track it later today (will come back to here when I do so and refer to it)

File: test/e2e/clusterctl_upgrade_test.go (hunk 10 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:40]: Absolutely!

No idea to be honest. But the ones who look at core CAPI for guidance might still do.

File: test/e2e/clusterctl_upgrade.go (hunk 5 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade.go]: Sounds good. Feel free to resolve then

**Comment by sbueringer** [comment]
Green :tada:

File: test/e2e/clusterctl_upgrade_test.go (hunk 10 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade_test.go:40]: Done

File: test/e2e/clusterctl_upgrade.go (hunk 5 above)
Comment (furkatgofurov7) [reply test/e2e/clusterctl_upgrade.go]: Coming back to this again, issue to track: https://github.com/kubernetes-sigs/cluster-api/issues/8535

**Review by killianmuldoon — COMMENTED** [review]
Looking good! Thanks for getting this one over the line - was a bit more complex than usual 🙂 

/lgtm

**Comment by furkatgofurov7** [comment]
@killianmuldoon @sbueringer thx folks, do we need to back-port it to release-1.4 branch?

**Comment by sbueringer** [comment]
> @killianmuldoon @sbueringer thx folks, do we need to back-port it to release-1.4 branch?

Yes, please!

I think we can & should definitely prepare already. Then we have to decide if we want to "officially" support 1.27 already in 1.4.2 or merge it intentionally afterwards to get more time in CI

**Comment by furkatgofurov7** [comment]
> Yes, please!
> 
> I think we can & should definitely prepare the PR already. Then we have to decide if we want to "officially" support 1.27 already in 1.4.2 or merge it intentionally afterwards to get more time in CI

ack, let me try the cherry-picker and see if it can do it otherwise will manually cherry pick it:

/cherry-pick release-1.4

**Comment by furkatgofurov7** [comment]
> No chance with the migration doc :D

Here: https://github.com/kubernetes-sigs/cluster-api/pull/8538

**Comment by johannesfrey** [comment]
/area provider/infrastructure-docker e2e-testing documentation
